src/server/snapshot.cc

```diff
@@ -61,6 +61,35 @@ void SliceSnapshot::Start(bool stream_journal, const Cancellation* cll) {
   });
 }
 
+void SliceSnapshot::StartIncremental(const Cancellation* cll, LSN start_lsn) {
+  auto* journal = db_slice_->shard_owner()->journal();
+  DCHECK(journal);
+
+  serializer_ = std::make_unique<RdbSerializer>(compression_mode_);
+
+  snapshot_fb_ =
+      fb2::Fiber("incremental_snapshot", [this, journal, cll, lsn = start_lsn]() mutable {
+        // The LSN we get from the replica is the last entry that
+        // was processed, so no need to send it.
+        lsn++;
+        while (!cll->IsCancelled() && journal->IsLSNInBuffer(lsn)) {
+          // TODO: Should we send multiple entries together?
+          serializer_->WriteJournalEntry(journal->GetEntry(lsn));
+          PushSerializedToChannel(false);
+          lsn++;
+        }
+        if (journal->GetLsn() == (lsn - 1)) {
+          serializer_->SendFullSyncCut();
+          PushSerializedToChannel(true);
```

Comment:
I think that we can preempt here and this can cause loosing writing entries to journal as you call register after this line